python/mxnet/contrib/quantization.py

```diff
@@ -543,3 +546,249 @@ def quantize_model(sym, arg_params, aux_params,
+        logger.info('Calibrating quantized symbol')
+        qsym = _calibrate_quantized_sym(qsym, th_dict)
+    else:
+        raise ValueError('please set calibration mode to naive or entropy.')
+
+    logger.info('Quantizing parameters')
+    qarg_params = _quantize_params(qsym, arg_params, th_dict)
+
+    return qsym, qarg_params, aux_params
+
+def static_net_forward(sym, arg_params, aux_params, collector,
+                       calib_data, data_shapes, ctx=cpu(), logger=logging):
+    """symbolic forward a static model with gluon dataset.
+    """
+    num_batches = len(calib_data)
+    num_inputs = len(data_shapes)
+    data_names = ()
+    data_shapes_ = []
+    if num_inputs == 1:
+        data_names = ('data',)
+        data_shapes_ = [(data_names[0], data_shapes[0])]
+    elif num_inputs > 1:
+        for i in range(num_inputs):
+            data_names = data_names + ('data' + str(i),)
+            data_shapes_.append((data_names[i], data_shapes[i]))
+    else:
+        raise ValueError('symbol must have at least one inputs.')
+    mod = Module(symbol=sym, context=ctx,
+                 data_names=data_names, label_names=None)
+    mod.bind(for_training=False, data_shapes=data_shapes_)
+    mod.set_params(arg_params, aux_params,
+                   allow_missing=False, force_init=True)
+    mod._exec_group.execs[0].set_monitor_callback(
+        collector.collect, monitor_all=True)
+    for batch in calib_data:
+        mod.forward(data_batch=batch, is_train=False)
+    if logger is not None:
+        logger.info("Collected statistics from %d batches"
+                    % num_batches)
+    return collector, data_names
+
+def save_params(fname, arg_params, aux_params, logger=logging):
+    save_dict = {('arg:%s' % k): v.as_in_context(cpu())
+                 for k, v in arg_params.items()}
+    save_dict.update({('aux:%s' % k): v.as_in_context(cpu())
+                      for k, v in aux_params.items()})
+    nd_save(fname, save_dict)
+
+def quantize_net(network, quantized_dtype='auto', exclude_layers=None, calib_data=None,
+                 data_shapes=None, calib_mode='none', ctx=cpu(), logger=logging):
+    """quantize a gluon net
+    """
+
+    logger.info('Export HybridBlock')
+    network.hybridize()
+    import mxnet as mx
+    data_nd = []
+    for shape in data_shapes:
+        data_nd.append(mx.nd.zeros(shape))
+    network(*data_nd)
+    import tempfile
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        prefix = os.path.join(tmpdirname, 'tmp')
+        network.export(prefix, epoch=0)
+        symnet, args, auxs = mx.model.load_checkpoint(prefix, 0)
+    logger.info('Exclude all FullyConnect and Flatten layers.')
+    if exclude_layers is None:
+        exclude_layers = []
+    for layers in list(symnet.get_internals()):
+        if layers.name.find('dense') != -1 or layers.name.find('flatten') != -1:
+            exclude_layers.append(layers.name)
+    symnet = symnet.get_backend_symbol('MKLDNN')
```

Comment:
Please use MKLDNN_QUANTIZE here too, to avoid some fusion that only appropriate to FP32 flow.

Reply:
done